### PR TITLE
Fix LoRA with grouped queries

### LINKED ISSUE #348: Dimension error with Falcon-7B and LoRA
Hello,

I am encountering an error when trying to finetune Falcon-7B on two Nvidia Titan V GPUs. I am giving this command:

```bash
python finetune/lora.py --checkpoint_dir checkpoints/tiiuae/falcon-7b-instruct/ --data_dir data/alpaca/ --precision 16-mixed
```

and I receive the following error:

```
RuntimeError: Given groups=3, expected weight to be divisible by 3 at dimension 0, but got weight of size [[4672, 2, 1]] instead
```

I have modified a few parameters in `finetune/lora.py` in order to avoid OOM errors.

```python
eval_interval = 100
save_interval = 100
eval_iters = 100
log_interval = 1
devices = 2  # MODIFIED 1 -> 2
# change this value to force a maximum sequence length
override_max_seq_length = None

# Hyperparameters
learning_rate = 3e-4
batch_size = 1  # MODIFIED 128 -> 1
micro_batch_size = 1  # MODIFIED 4 -> 1
gradient_accumulation_iters = batch_size // micro_batch_size
assert gradient_accumulation_iters > 0
max_iters = 50000  # train dataset size
weight_decay = 0.01
lora_r = 2  # MODIFIED 8 -> 2
lora_alpha = 16
lora_dropout = 0.05
lora_query = True
lora_key = True  # MODIFIED False -> True
lora_value = True
lora_projection = False
lora_mlp = False
lora_head = False
warmup_steps = 100
...
```

The error seems to only appear after setting `lora_key=True` . As I understand it, this should apply a low rank approximation to the keys weight matrix for each attention head. Is this understanding correct and if so, does anyone know why this error is occuring?

Thank you for your time!

PS - here is the full out put including stack trace:

```
python finetune/lora.py --checkpoint_dir checkpoints/tiiuae/falcon-7b-instruct/ --data_dir data/alpaca/ --precision 16-mixed
Initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/2
Initializing distributed: GLOBAL_RANK: 1, MEMBER: 2/2
----------------------------------------------------------------------------------------------------
distributed_backend=nccl
All distributed processes registered. Starting with 2 processes
----------------------------------------------------------------------------------------------------

{'eval_interval': 100, 'save_interval': 100, 'eval_iters': 100, 'log_interval': 1, 'devices': 2, 'learning_rate': 0.0003, 'batch_size': 1, 'micro_batch_size': 1, 'gradient_accumulation_iters': 1, 'max_iters': 50000, 'weight_decay': 0.01, 'lora_r': 2, 'lora_alpha': 16, 'lora_dropout': 0.05, 'lora_query': True, 'lora_key': True, 'lora_value': True, 'lora_projection': False, 'lora_mlp': False, 'lora_head': False, 'warmup_steps': 100}
[rank: 1] Global seed set to 1337
[rank: 0] Global seed set to 1337
Loading model 'checkpoints/tiiuae/falcon-7b-instruct/lit_model.pth' with {'org': 'tiiuae', 'name': 'falcon-7b-instruct', 'block_size': 2048, 'vocab_size': 50254, 'padding_multiple': 512, 'padded_vocab_size': 65024, 'n_layer': 32, 'n_head': 71, 'n_embd': 4544, 'rotary_percentage': 1.0, 'parallel_residual': True, 'bias': False, 'n_query_groups': 1, 'shared_attention_norm': True, '_norm_class': 'LayerNorm', 'norm_eps': 1e-05, '_mlp_class': 'GptNeoxMLP', 'intermediate_size': 18176, 'condense_ratio': 1, 'r': 2, 'alpha': 16, 'dropout': 0.05, 'to_query': True, 'to_key': True, 'to_value': True, 'to_projection': False, 'to_mlp': False, 'to_head': False}
Number of trainable parameters: 1,171,456
Number of non trainable parameters: 7,217,189,760
[rank: 1] Global seed set to 1338
[rank: 0] Global seed set to 1337
Validating ...
Traceback (most recent call last):
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/finetune/lora.py", line 319, in <module>
    CLI(setup)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/jsonargparse/_cli.py", line 85, in CLI
    return _run_component(component, cfg_init)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/jsonargparse/_cli.py", line 147, in _run_component
    return component(**cfg)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/finetune/lora.py", line 81, in setup
    fabric.launch(main, data_dir, checkpoint_dir, out_dir)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/lightning/fabric/fabric.py", line 806, in launch
    return self._wrap_and_launch(function, self, *args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/lightning/fabric/fabric.py", line 888, in _wrap_and_launch
    return to_run(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/lightning/fabric/fabric.py", line 893, in _wrap_with_setup
    return to_run(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/finetune/lora.py", line 135, in main
    train(fabric, model, optimizer, train_data, val_data, checkpoint_dir, out_dir, speed_monitor)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/finetune/lora.py", line 156, in train
    validate(fabric, model, val_data, tokenizer, longest_seq_length)  # sanity check
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/finetune/lora.py", line 241, in validate
    logits = model(input_ids)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
Traceback (most recent call last):
  File "finetune/lora.py", line 319, in <module>
    return self._call_impl(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    CLI(setup)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/jsonargparse/_cli.py", line 85, in CLI
    return _run_component(component, cfg_init)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/jsonargparse/_cli.py", line 147, in _run_component
    return component(**cfg)
  File "finetune/lora.py", line 81, in setup
    return forward_call(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/lightning/fabric/wrappers.py", line 116, in forward
    fabric.launch(main, data_dir, checkpoint_dir, out_dir)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/lightning/fabric/fabric.py", line 806, in launch
    output = self._forward_module(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._wrap_and_launch(function, self, *args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/lightning/fabric/fabric.py", line 887, in _wrap_and_launch
    return self._call_impl(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return launcher.launch(to_run, *args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/lightning/fabric/strategies/launchers/subprocess_script.py", line 90, in launch
    return function(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/lightning/fabric/fabric.py", line 893, in _wrap_with_setup
    return forward_call(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py", line 821, in forward
    return to_run(*args, **kwargs)
  File "finetune/lora.py", line 135, in main
    output = self._fsdp_wrapped_module(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    train(fabric, model, optimizer, train_data, val_data, checkpoint_dir, out_dir, speed_monitor)
  File "finetune/lora.py", line 156, in train
    validate(fabric, model, val_data, tokenizer, longest_seq_length)  # sanity check
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "finetune/lora.py", line 241, in validate
    logits = model(input_ids)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
    return self._call_impl(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/lit_gpt/lora.py", line 516, in forward
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    x, *_ = block(x, (cos, sin), max_seq_length)
    return forward_call(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/lightning/fabric/wrappers.py", line 116, in forward
    output = self._forward_module(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py", line 821, in forward
    return forward_call(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py", line 821, in forward
    output = self._fsdp_wrapped_module(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    output = self._fsdp_wrapped_module(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py", line 164, in forward
    return self.checkpoint_fn(  # type: ignore[misc]
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/_compile.py", line 24, in inner
    return forward_call(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/lit_gpt/lora.py", line 516, in forward
    return torch._dynamo.disable(fn, recursive)(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py", line 321, in _fn
    return fn(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/_dynamo/external_utils.py", line 17, in inner
    return fn(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/utils/checkpoint.py", line 457, in checkpoint
    x, *_ = block(x, (cos, sin), max_seq_length)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    ret = function(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py", line 821, in forward
    return forward_call(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/lit_gpt/model.py", line 168, in forward
    h, new_kv_cache = self.attn(n_1, rope, max_seq_length, mask, input_pos, kv_cache)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    output = self._fsdp_wrapped_module(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/lit_gpt/model.py", line 205, in forward
    qkv = self.attn(x)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return forward_call(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py", line 164, in forward
    return self.checkpoint_fn(  # type: ignore[misc]
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/_compile.py", line 24, in inner
    return torch._dynamo.disable(fn, recursive)(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py", line 321, in _fn
    return fn(*args, **kwargs)
    return self._call_impl(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/_dynamo/external_utils.py", line 17, in inner
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return fn(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/utils/checkpoint.py", line 457, in checkpoint
    ret = function(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return forward_call(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/lit_gpt/lora.py", line 369, in forward
    return self._call_impl(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    after_B = F.conv1d(
RuntimeError: Given groups=3, expected weight to be divisible by 3 at dimension 0, but got weight of size [[4672, 2, 1]] instead
    return forward_call(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/lit_gpt/model.py", line 168, in forward
    h, new_kv_cache = self.attn(n_1, rope, max_seq_length, mask, input_pos, kv_cache)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/lit_gpt/model.py", line 205, in forward
    qkv = self.attn(x)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/home/sam.voisin/healthcare-data-suite/lit-gpt/lit_gpt/lora.py", line 369, in forward
    after_B = F.conv1d(
RuntimeError: Given groups=3, expected weight to be divisible by 3 at dimension 0, but got weight of size [[4672, 2, 1]] instead
```

### LINKED ISSUE #357: LoRA with query groups doesn't work correctly
This is a more technical analysis of this issue: https://github.com/Lightning-AI/lit-gpt/issues/348 (check it to the see the error output)

TL;DR: `groups` in `nn.Conv1d` splits tensor in an equal parts and they are processed independently, but if the number of query groups is smaller than the number of heads then in a combined matrix `QKV` key and value parts will be smaller than query one and thus will be processed incorrectly by `nn.Conv1d`. 

---

In `LoRAQKVLinear` we have a combined matrix QKV that consists of three matrices: query, key and value. So the output dimension of the matrix consists of three terms: number of heads * head size * 3 (because all three matrices are of equal size). 
But with [query groups](https://github.com/Lightning-AI/lit-gpt/blob/e40647fd7ed4a9b5c8634905546835757eb7c9b7/lit_gpt/config.py#L29) the actual dimension will be: 
1. query: `number of heads` * `head size`
2. key: `number of query groups` * `head_size`
3. value: `number of query groups` * `head_size`

So if the `number of query groups` < `number of heads`, then the portion of key and value in the combined matrix will be smaller than the portion of the query. In other words QKV will consist of parts of unequal size.
And the problem is in how `nn.Conv1d` works with groups parameter. For instance for groups=2 from the official documentation:
> At groups=2, the operation becomes equivalent to having two conv layers side by side, each seeing half the input channels and producing half the output channels, and both subsequently concatenated.

Which means that no matter what the number of groups is, the tensor will be divided in equal parts and each part is processed independently. But as I stated above the tensor might consist of unequal sized parts.

The issue was overlooked as the number of groups was 2 (`enable_lora` was hardcoded as (True, False, True)) and the dimension of [`after_A`](https://github.com/Lightning-AI/lit-gpt/blob/e40647fd7ed4a9b5c8634905546835757eb7c9b7/lit_gpt/lora.py#L370) is even thus divisible by two and thus no errors are raised but in fact it is incorrect. With the number of groups of 3 (LoRA is applied to all three matrices in QKV) we get the error.

---

Ideally would be to have an ability to provide a custom way to split tensor in `nn.Conv1d`, but it seems to be impossible.
So the only option is to do everything manually (if number of query groups != number of heads):
1. split `after_A` tensor
2. split `lora_B` weight matrix
3. apply independently
4. concatenate the result

But of course it's not particularly performant.
Any ideas?

### PR DESCRIPTION
**Pull request by Andrei-Aksionov**
Hi there 👋 

Fixes #348: there you can find error output
Fixes #357:  a technical explanation of the problem

----

In a nutshell the solution is to use `torch.nn.functional.conv1d` with `groups` parameter when grouped queries are not used (number of heads is equal to the number of query groups: see the scheme in `lit_gpt/config.py:Config`) and in other case do it manually: split input and weight tensors to the correct shapes, apply each weight matrix to the corresponding input matrix, concatenate the result.

----

Tested on a single T4 with `Pythia-410m` with default parameters (except `max_iters` which I set to a 1,000).
```bash
python finetune/lora.py --checkpoint_dir checkpoints/EleutherAI/pythia-410m  --data_dir data/alpaca/ --precision 16-mixed
```

I used this model as it's relatively small and doesn't require these changes (as it doesn't use grouped queries) so it allows to compare two approaches: conv1d with `groups` parameter and manual one with splitting and concatenating tensors:
1. conv1d with groups finished in 169 seconds
2. manual approach finished in 128 seconds

Loss after 1,000 iterations is the same in both cases: ~1.801

So it looks like splitting a tensor is not only doesn't hurt performance, but in fact brings some speed improvements. At least in case of this particular model on this particular dataset running on this particular hardware.

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]

File: lit_gpt/lora.py [hunk 1]
```diff
@@ -208,8 +210,14 @@ def __init__(
...
             self.kv_embd_size = self.linear.in_features // (n_head // n_query_groups)
-            shape = self.linear.in_features * enable_q + self.kv_embd_size * enable_k + self.kv_embd_size * enable_v
-            self.lora_B = nn.Parameter(self.linear.weight.new_zeros(shape, r))  # (256, 2))
+            # qkv_shapes will be used to split a tensor with weights correctly
+            self.qkv_shapes = tuple(
+                filter(
+                    None,
+                    (self.linear.in_features * enable_q, self.kv_embd_size * enable_k, self.kv_embd_size * enable_v),
+                )
+            )
```
Comment (carmocca) [inline lit_gpt/lora.py:219]: I'd find this easier to read

```suggestion
            qkv_shapes = (self.linear.in_features * enable_q, self.kv_embd_size * enable_k, self.kv_embd_size * enable_v)
            self.qkv_shapes = [s for s in qkv_shapes if s is not None]
```

File: tests/test_lora.py [hunk 2]
```diff
@@ -261,6 +263,75 @@ def test_lora_layer_forward_no_exception(apply_to):
...
+@torch.inference_mode()
+@pytest.mark.parametrize(
+    ("n_head", "enable_lora"),
+    list(
+        product(
+            (1, 2, 3, 6, 12),
+            [p for p in product((False, True), repeat=3) if any(p)] # skip enable_lora=(False, False, False)
+        ),
+    ),
+)
```
Comment (carmocca) [inline tests/test_lora.py:310]:
```suggestion
@pytest.mark.parametrize("n_head", (1, 2, 3, 6, 12))
@pytest.mark.parametrize(
    "enable_lora",
    [
        (False, False, True),
        (False, True, False),
        (False, True, True),
        (True, False, False),
        (True, False, True),
        (True, True, False),
        (True, True, True),
    ],
)
```

**Review by carmocca — COMMENTED** [review]
> conv1d with groups finished in 169 seconds
manual approach finished in 128 seconds

I'm confused by these numbers. If manual is faster, why wouldn't we always do it?

Can you benchmark the conv1d calls specifically? You can use the code [here](https://gist.github.com/Chillee/f86675147366a7a0c6e244eaa78660f7) (ignore the torch.compile bits)

File: tests/test_lora.py (hunk 2 above)
Comment (Andrei-Aksionov) [reply tests/test_lora.py]: Didn't know that it's possible. 
Thanks!

**Comment by Andrei-Aksionov** [comment]
> I'm confused by these numbers. If manual is faster, why wouldn't we always do it?

a) it's not thoroughly tested
b) this PR tackles problems with LoRA applied when grouped queries are used


> Can you benchmark the conv1d calls specifically? You can use the code [here](https://gist.github.com/Chillee/f86675147366a7a0c6e244eaa78660f7) (ignore the torch.compile bits)

When worked on this PR I've noticed that conv1d call itself is slower and [tests](https://drive.google.com/file/d/1pBWoHDih34eZh6Fq7XkBukWs4w9TlEDe/view?usp=sharing) reflect this.
The speedup comes apparently from backward pass / optimization steps maybe 🤷 .
Though it needs to be tested: I tried the smallest Pythia-70m with 1k iterations and:
- conv1d with groups finished in 44.01s
- manual approach finished in 41.22s

the loss is the same.

So maybe I'll test it more vigorously and on various models and if there is a stable speed improvement - I'll create a PR.
